### PR TITLE
Serialization/Deserialization of bitcode to std::ostream/std::istream

### PR DESCRIPTION
**Pull request by jmmartinez**

### CODE REVIEW

File: cpplib/include/easy/runtime/BitcodeTracker.h [hunk 1]
```diff
@@ -30,13 +30,16 @@ class BitcodeTracker {
...
+  std::unordered_map<std::string, void*> NameToAddress;
 
   public:
 
   void registerFunction(void* FPtr, const char* Name, GlobalMapping* Globals, const char* Bitcode, size_t BitcodeLen) {
     Functions.emplace(FPtr, FunctionInfo{Name, Globals, Bitcode, BitcodeLen});
+    NameToAddress.emplace(Name, FPtr);
   }
 
+  void* getNameAddress(std::string const &Name);
```
Comment (serge-sans-paille) [inline cpplib/include/easy/runtime/BitcodeTracker.h]: s/getNameAddress/getAddress/ ?

File: cpplib/include/easy/jit.h [hunk 2]
```diff
@@ -38,6 +38,14 @@ class FunctionWrapperBase {
   void* getRawPointer() const {
```
Comment (serge-sans-paille) [inline cpplib/include/easy/jit.h]: missing ``#include <iostream>`` somewhere?

File: runtime/Function.cpp [hunk 3]
```diff
@@ -110,12 +132,42 @@ std::unique_ptr<Function> Function::Compile(void *Addr, easy::Context const& C)
...
+  }
+
+  auto M = std::move(ModuleOrError.get());
+
+  std::string FunName = easy::GetEntryFunctionName(*M);
+
+  GlobalMapping* Globals = nullptr;
+  if(void* OrigFunPtr = BT.getNameAddress(FunName)) {
+    const char* Name;
+    std::tie(Name, Globals) = BT.getNameAndGlobalMapping(OrigFunPtr);
```
Comment (serge-sans-paille) [inline runtime/Function.cpp]: ``Name`` sould be ``std::ignore``

File: runtime/Utils.cpp [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+    MDNode const* MD = F.getMetadata("easy::jit");
+    if(!MD && !MD->getNumOperands())
+      continue;
+    Metadata const* Operand = MD->getOperand(0);
+
+    MDString const *String = dyn_cast<MDString const>(Operand);
+    if(String->getString() == "entry")
+      return F.getName();
+  }
+  return "";
```
Comment (serge-sans-paille) [inline runtime/Utils.cpp]: $\mathcal{0}(|F|)$ but is there a better way?

File: runtime/pass/InlineParameters.cpp [hunk 5]
```diff
@@ -173,6 +173,12 @@ bool easy::InlineParameters::runOnModule(llvm::Module &M) {
   for(Attribute Attr : FunAttrs)
     WrapperFun->addFnAttr(Attr);
 
+  // add metadata to identify the entry function
+  LLVMContext &Ctx = M.getContext();
+  Metadata* MDStr = MDString::get(Ctx, "entry");
+  MDNode* MD = MDNode::get(Ctx, {MDStr});
+  WrapperFun->addMetadata("easy::jit", *MD);
```
Comment (serge-sans-paille) [inline runtime/pass/InlineParameters.cpp]: the metadata name should be in a variable somewhere :-)

File: tests/simple/serialize_multifile.cpp [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+static int var = 0;
+
+static int add (int a, int b) {
+  return a+b+(var++);
+}
+
+std::string get_add(int b) {
+  auto inc_store = easy::jit(add, _1, 1);
+
+  std::stringstream out;
```
Comment (serge-sans-paille) [inline tests/simple/serialize_multifile.cpp]: could be ``ostringstream`

File: runtime/Function.cpp (hunk 3 above)
Comment (jmmartinez) [reply runtime/Function.cpp]: Done!

File: cpplib/include/easy/jit.h (hunk 2 above)
Comment (jmmartinez) [reply cpplib/include/easy/jit.h]: Done!

File: cpplib/include/easy/runtime/BitcodeTracker.h (hunk 1 above)
Comment (jmmartinez) [reply cpplib/include/easy/runtime/BitcodeTracker.h]: Done!

File: runtime/Utils.cpp (hunk 4 above)
Comment (jmmartinez) [reply runtime/Utils.cpp]: I will try using the named metadata.

**Pull request merged**